snuba/clickhouse/formatter.py

```diff
@@ -0,0 +1,95 @@
+from typing import Optional, Sequence
+from snuba.query.expressions import (
+    Column,
+    CurriedFunctionCall,
+    Expression,
+    ExpressionVisitor,
+    FunctionCall,
+    Literal,
+)
+from snuba.query.parsing import ParsingContext
+from snuba.util import escape_alias, escape_col, escape_string
+
+# Tokens used when formatting. Defining them as costant
+# will make it easy (if/when needed) to make some changes
+# to the layout of the output.
+NULL = "NULL"
+COMMA = ","
+SPACE = " "
+TRUE = "true"
+FALSE = "false"
+QUOTE = "'"
+OPEN_BRACKET = "("
+CLOSED_BRACKET = ")"
+DOT = "."
+
+
+class ClickhouseExpressionFormatter(ExpressionVisitor[str]):
+    """
+    This Visitor implementation is able to format one expression in the Snuba
+    Query for Clickhouse.
+
+    The only state mainained is the PrasingContext, which allows us to resolve
+    aliases and can be reused when formatting multiple expressions.
+
+    When passing an instance of this class to the accept method of
+    the visited expression, the return value is the formatted string.
+
+    This Formatter produces a properly escaped string. The result should never
+    be further escaped. This should be the only place where expression
+    escaping happens as it is done by each method that formats a specific
+    type of expression.
+    """
+
+    def __init__(self, parsing_context: Optional[ParsingContext] = None) -> None:
+        self.__parsing_context = parsing_context or ParsingContext()
+
+    def __alias(self, formatted_exp: str, alias: Optional[str]) -> str:
+        if not alias:
+            return formatted_exp
+        elif self.__parsing_context.is_alias_present(alias):
+            ret = escape_alias(alias)
+            # This is for the type checker. escape_alias can return None if
+            # we pass None. But here we do not pass None so a None return value
+            # is not valid.
+            assert ret is not None
+            return ret
+        else:
+            self.__parsing_context.add_alias(alias)
+            return f"{OPEN_BRACKET}{formatted_exp}{SPACE}AS{SPACE}{escape_alias(alias)}{CLOSED_BRACKET}"
+
+    def visitLiteral(self, exp: Literal) -> str:
+        if exp.value is None:
+            return NULL
+        elif exp.value is True:
+            return TRUE
+        elif exp.value is False:
+            return FALSE
+        elif isinstance(exp.value, str):
+            return escape_string(exp.value)
+        elif isinstance(exp.value, (int, float)):
+            return str(exp.value)
+        else:
+            raise ValueError(f"Invalid literal type {type(exp.value)}")
```

Comment:
It's possible that the possible types for `Literal.value` can change, this is probably better worded as:

```suggestion
            raise ValueError(f"Unexpected literal type: {type(exp.value)}")
```